src/compiler/checker.ts

```diff
@@ -1021,7 +1021,11 @@ namespace ts {
         function mergeSymbol(target: Symbol, source: Symbol, unidirectional = false): Symbol {
             if (!(target.flags & getExcludedSymbolFlags(source.flags)) ||
                 (source.flags | target.flags) & SymbolFlags.Assignment) {
-                Debug.assert(source !== target);
+                if (source === target) {
+                    // This can happen when an export assigned namespace exports something also erroneously exported at the top level

```

Comment:
is there any way to avoid this upstream? I *really* want to keep this assert around to prevent other incorrect cases.

Reply:
Nope, not really. In
```
export = exports;
declare class exports {
    constructor(p: number);
    t: number;
}
export class Sub {
    instance!: {
        t: number;
    };
}
declare namespace exports {
    export { Sub };
}
```
since there's an `export =`, the `export` modifier on `Sub` is erroneous, however it still gets registered into the `export`s of the `SourceFile`, alongside the `export=`. This then triggers the commonjs merge-down case (which triggers any time there's more than just an `export=` in a source file with an `export=`), and we end up needing to merge the (target of the) exported alias to `Sub` with `Sub` itself. That obviously doesn't do anything.

I just so happened to discover this during iteration on the PR (when I had a bug where I was adding too many export modifiers in my generated declarations) where I crashed the compiler on the assertion when I parsed and checked the output. Crashing on input is never the right choice for us, so correctly choosing a behavior here is better, now that we have a meaningful scenario under which we can choose that behavior (which, IMO, if a symbol tries to merge with itself, returning itself makes sense).

Reply:
Is this JS? I thought the merge didn't happen with JS. Or can you repro this with `module.exports = exports` in a JS file?

I'm still wary of this change, since we'll now be covering up errors; and ones that result from some inconsistent state that we haven't thought of. But if you want to keep it like this: callers of mergeSymbol can  currently assume that target is Transient. It's not anymore. We need to make sure that all the callers are OK with a non-transient symbol.

Reply:
no @sandersn - the above code is a declaration file, no JS flagging needed.

Reply:
The point, if I may be so blunt, is that without this change, the above code crashed the compiler. That is bad. Everything leading up to the assertion (including the speculative merge) is a reasonable action, given the error in the file, ergo the assertion must be unreasonable (as it does not admit code with errors, which does exist).